# Lab 8: Define and Solve an ML Problem of Your Choosing

In [17]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

# packages
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2" # suppress info and warning messages
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import tensorflow.keras as keras
import time

In this lab assignment, you will follow the machine learning life cycle and implement a model to solve a machine learning problem of your choosing. You will select a data set and choose a predictive problem that the data set supports.  You will then inspect the data with your problem in mind and begin to formulate a  project plan. You will then implement the machine learning project plan. 

You will complete the following tasks:

1. Build Your DataFrame
2. Define Your ML Problem
3. Perform exploratory data analysis to understand your data.
4. Define Your Project Plan
5. Implement Your Project Plan:
    * Prepare your data for your model.
    * Fit your model to the training data and evaluate your model.
    * Improve your model's performance.

## Part 1: Build Your DataFrame

You will have the option to choose one of four data sets that you have worked with in this program:

* The "census" data set that contains Census information from 1994: `censusData.csv`
* Airbnb NYC "listings" data set: `airbnbListingsData.csv`
* World Happiness Report (WHR) data set: `WHR2018Chapter2OnlineData.csv`
* Book Review data set: `bookReviewsData.csv`

Note that these are variations of the data sets that you have worked with in this program. For example, some do not include some of the preprocessing necessary for specific models. 

#### Load a Data Set and Save it as a Pandas DataFrame

The code cell below contains filenames (path + filename) for each of the four data sets available to you.

<b>Task:</b> In the code cell below, use the same method you have been using to load the data using `pd.read_csv()` and save it to DataFrame `df`. 

You can load each file as a new DataFrame to inspect the data before choosing your data set.

In [18]:
# File names of the four data sets
adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")
airbnbDataSet_filename = os.path.join(os.getcwd(), "data", "airbnbListingsData.csv")
WHRDataSet_filename = os.path.join(os.getcwd(), "data", "WHR2018Chapter2OnlineData.csv")
bookReviewDataSet_filename = os.path.join(os.getcwd(), "data", "bookReviewsData.csv")


df = pd.read_csv(bookReviewDataSet_filename, header=0)

df.head()

Review  Positive Review
0  This was perhaps the best of Johannes Steinhof...             True
1  This very fascinating book is a story written ...             True
2  The four tales in this collection are beautifu...             True
3  The book contained more profanity than I expec...            False
4  We have now entered a second time of deep conc...             True

## Part 2: Define Your ML Problem

Next you will formulate your ML Problem. In the markdown cell below, answer the following questions:

1. List the data set you have chosen.
2. What will you be predicting? What is the label?
3. Is this a supervised or unsupervised learning problem? Is this a clustering, classification or regression problem? Is it a binary classificaiton or multi-class classifiction problem?
4. What are your features? (note: this list may change after your explore your data)
5. Explain why this is an important problem. In other words, how would a company create value with a model that predicts this label?

1. Book Review dataset 
2. I will be predicting whether the book review has a positive or negative connotation. The label is true/false, true denoting that it is a positive review, and false denoting that it is a negative review. 
3. This is a supervised learning problem. This is a classification problem, specifically a binary classification problem.
4. The features are the text of the book review which will be converted in TF-IDF vectors, giving us numerical features. 
5. This is an important problem because this can further inform companies on how books may perform and be perceived by readers for future publications. This can better inform what books are liked by readers, and what features of these books make them positively perceived. 

## Part 3: Understand Your Data

The next step is to perform exploratory data analysis. Inspect and analyze your data set with your machine learning problem in mind. Consider the following as you inspect your data:

1. What data preparation techniques would you like to use? These data preparation techniques may include:

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    

2. What machine learning model (or models) you would like to use that is suitable for your predictive problem and data?
    * Are there other data preparation techniques that you will need to apply to build a balanced modeling data set for your problem and model? For example, will you need to scale your data?
 
 
3. How will you evaluate and improve the model's performance?
    * Are there specific evaluation metrics and methods that are appropriate for your model?
    

Think of the different techniques you have used to inspect and analyze your data in this course. These include using Pandas to apply data filters, using the Pandas `describe()` method to get insight into key statistics for each column, using the Pandas `dtypes` property to inspect the data type of each column, and using Matplotlib and Seaborn to detect outliers and visualize relationships between features and labels. If you are working on a classification problem, use techniques you have learned to determine if there is class imbalance.

<b>Task</b>: Use the techniques you have learned in this course to inspect and analyze your data. You can import additional packages that you have used in this course that you will need to perform this task.

<b>Note</b>: You can add code cells if needed by going to the <b>Insert</b> menu and clicking on <b>Insert Cell Below</b> in the drop-drown menu.

In [19]:
# missing reviews 
missing_reviews = df['Review'].isnull().sum()
print("Missing Reviews:", missing_reviews)

# missing labels 
missing_labels = df['Positive Review'].isnull().sum()
print("Missing Label:", missing_labels)

# check class imbalance
imbalance = df['Positive Review'].value_counts()
print(imbalance)
# no detected imbalances

# labeled examples 
y = df['Positive Review']      # label
X = df['Review']               # feature


# training/test data sets 
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, random_state=1234)

# create and fit vectorizer 
tfidf_vectorizer = TfidfVectorizer(stop_words='english',
                                   min_df=5, 
                                   ngram_range=(1,1), 
                                   max_features = 5000)
     # NOTE: auto
tfidf_vectorizer.fit(X_train)
# transform training and test data 
X_train_tfidf = tfidf_vectorizer.transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)


print("X_train_tfidf shape:", X_train_tfidf.shape)
print("X_test_tfidf shape:", X_test_tfidf.shape)
#print("y_train shape:", y_train.shape)
#print("y_test shape:", y_test.shape)
#print("y_proba shape:", y_proba.shape)

      
vocab_size = len(tfidf_vectorizer.vocabulary_)
print("Vocab Size:", vocab_size)

Missing Reviews: 0
Missing Label: 0
False    993
True     980
Name: Positive Review, dtype: int64
X_train_tfidf shape: (1479, 3583)
X_test_tfidf shape: (494, 3583)
Vocab Size: 3583


## Part 4: Define Your Project Plan

Now that you understand your data, in the markdown cell below, define your plan to implement the remaining phases of the machine learning life cycle (data preparation, modeling, evaluation) to solve your ML problem. Answer the following questions:

* Do you have a new feature list? If so, what are the features that you chose to keep and remove after inspecting the data? 
* Explain different data preparation techniques that you will use to prepare your data for modeling.
* What is your model (or models)?
* Describe your plan to train your model, analyze its performance and then improve the model. That is, describe your model building, validation and selection plan to produce a model that generalizes well to new data. 

1. I chose to keep all the features, since there was no noticeable class imbalance among the labels such that there were 993 non-positive reviews and 980 positive reviews.
2. I moved forward with transforming the text into numerical vectors as a method of data preparation. 
3. My model is a feedforward neural network
4. I plan to construct a neural network, testing different number of hidden layers with different units where each layer will use the ReLU activation function. But first I would create a sgd optimizer, a binary cross entropy function, and a loss function, then compile the model which would act on my split validation and test sets. I would analyze performance using training and accuracy loss metrics and a ROC-AUC score.
5. 

## Part 5: Implement Your Project Plan

<b>Task:</b> In the code cell below, import additional packages that you have used in this course that you will need to implement your project plan.

In [20]:
# packages
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2" # suppress info and warning messages
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import tensorflow.keras as keras
import time


<b>Task:</b> Use the rest of this notebook to carry out your project plan. 

You will:

1. Prepare your data for your model.
2. Fit your model to the training data and evaluate your model.
3. Improve your model's performance by performing model selection and/or feature selection techniques to find best model for your problem.

Add code cells below and populate the notebook with commentary, code, analyses, results, and figures as you see fit. 

In [21]:
# CONSTRUCT NEURAL NETWORK

# model structure 
nn_model = keras.Sequential()
inputLayer = keras.layers.InputLayer(input_shape=(vocab_size,), name='input')
nn_model.add(inputLayer)
nn_model.add(keras.layers.Dropout(0.2))

# HIDDEN LAYERS 
hidden_1 = keras.layers.Dense(units=64, activation='relu', name='h1_1')
nn_model.add(hidden_1)
nn_model.add(keras.layers.Dropout(0.2))

hidden_2 = keras.layers.Dense(units=32, activation='relu', name='h1_2')
nn_model.add(hidden_2)
nn_model.add(keras.layers.Dropout(0.2))

hidden_3 = keras.layers.Dense(units=16, activation='relu', name='h1_3')
nn_model.add(hidden_3)
nn_model.add(keras.layers.Dropout(0.2))


outputLayer = keras.layers.Dense(units=1, activation='sigmoid', name='output')
nn_model.summary()   # summary 

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dropout (Dropout)            (None, 3583)              0         
_________________________________________________________________
h1_1 (Dense)                 (None, 64)                229376    
_________________________________________________________________
dropout_1 (Dropout)          (None, 64)                0         
_________________________________________________________________
h1_2 (Dense)                 (None, 32)                2080      
_________________________________________________________________
dropout_2 (Dropout)          (None, 32)                0         
_________________________________________________________________
h1_3 (Dense)                 (None, 16)                528       
_________________________________________________________________
dropout_3 (Dropout)          (None, 16)               

In [22]:
# OPTIMIZATION FUNCTION
sgdOptimizer = keras.optimizers.SGD(learning_rate=0.01)
lossFN = keras.losses.BinaryCrossentropy(from_logits=False)
nn_model.compile(optimizer = sgdOptimizer,
                loss = lossFN,
                metrics=['accuracy'])

In [23]:
class ProgBarLoggerNEpochs(keras.callbacks.Callback):
    
    def __init__(self, num_epochs: int, every_n: int = 50):
        self.num_epochs = num_epochs
        self.every_n = every_n
    
    def on_epoch_end(self, epoch, logs=None):
        if (epoch + 1) % self.every_n == 0:
            s = 'Epoch [{}/ {}]'.format(epoch + 1, self.num_epochs)
            logs_s = ['{}: {:.4f}'.format(k.capitalize(), v)
                      for k, v in logs.items()]
            s_list = [s] + logs_s
            print(', '.join(s_list))


In [24]:

t0 = time.time() # start time

num_epochs = 32 # epochs

history = nn_model.fit(
    X_train_tfidf.toarray(),    # convert to a numpy array 
    y_train,
    epochs=num_epochs,
    batch_size=64,
    verbose=0,
    validation_split=0.20,
    callbacks=[ProgBarLoggerNEpochs(num_epochs, every_n=5)]
)

t1 = time.time() # stop time

print('Elapsed time: %.2fs' % (t1-t0))

Epoch [5/ 32], Loss: 2.7924, Accuracy: 0.0389, Val_loss: 1.6690, Val_accuracy: 0.0000
Epoch [10/ 32], Loss: 2.4169, Accuracy: 0.0541, Val_loss: 1.2008, Val_accuracy: 0.0000
Epoch [15/ 32], Loss: 2.4202, Accuracy: 0.0592, Val_loss: 1.1718, Val_accuracy: 0.0034
Epoch [20/ 32], Loss: 2.4011, Accuracy: 0.0634, Val_loss: 1.1591, Val_accuracy: 0.0034
Epoch [25/ 32], Loss: 2.4056, Accuracy: 0.0862, Val_loss: 1.1529, Val_accuracy: 0.0304
Epoch [30/ 32], Loss: 2.4323, Accuracy: 0.0786, Val_loss: 1.1493, Val_accuracy: 0.0845
Elapsed time: 2.17s


In [7]:
# PLOTTING 
plt.plot(range(1, num_epochs + 1), history.history['loss'], label='Training Loss')
plt.plot(range(1, num_epochs + 1), history.history['val_loss'], label='Validation Loss')

plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()
plt.show()


# Plot training and validation accuracy
plt.plot(range(1, num_epochs + 1), history.history['accuracy'], label='Training Accuracy')
plt.plot(range(1, num_epochs + 1), history.history['val_accuracy'], label='Validation Accuracy')

plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.legend()
plt.show()


NameError: name 'num_epochs' is not defined

In [8]:
loss, accuracy = nn_model.evaluate(X_test_tfidf.toarray(), y_test)

print('Loss: ', str(loss) , 'Accuracy: ', str(accuracy))

NameError: name 'nn_model' is not defined

In [9]:
# Make predictions on the test set

print("Predictions for the first 20 examples:")
# y_test_array = y_test.to_numpy()
y_proba = nn_model.predict(X_test_tfidf.toarray()).flatten()

# Prepare test labels
y_test_array = y_test.reset_index(drop=True).to_numpy().astype(int)

print("y_proba shape:", y_proba.shape)
print("y_test_array shape:", y_test_array.shape)


for i in range(30):
    prob = y_proba[i]
    predicted_class = 1 if prob > 0.5 else 0 
    actual_class = y_test_array[i]
    print(f"Example {i+1}: Probability = {prob:.4f}, Predicted = {predicted_class}, Actual = {actual_class}")


Predictions for the first 20 examples:


NameError: name 'nn_model' is not defined